libs/components/config/src/lib/params.ts

```diff
@@ -178,44 +178,103 @@ export class SkyAppRuntimeConfigParams {
-    const existingParams = getUrlSearchParams(url);
+  #buildUrlWithParams(
+    url: string,
+    excludeParams: Set<string>,
+    queryParams?: Map<string, string | number | boolean>
+  ): string {
+    const urlParams = getUrlSearchParams(url);
+
+    // Determine if the url params need to be modified.
+    const urlParamsRequireModification = urlParams
+      .keys()
+      .some((key) => excludeParams.has(key) || queryParams?.has(key));
+
+    const [beforeFragment, fragment] = url.split('#', 2);
+
+    const root = urlParamsRequireModification
+      ? beforeFragment.split('?')[0]
+      : beforeFragment;
+
+    let unifiedParams = urlParamsRequireModification
+      ? urlParams
+      : getUrlSearchParams('');
 
-    // When encoding the params, keep the requested params separate so that
-    // we can leave the existing params as-is.
-    let requestedParams = getUrlSearchParams('');
+    // Remove excluded params.
+    if (urlParamsRequireModification) {
+      for (const paramName of excludeParams.keys()) {
+        unifiedParams = unifiedParams.delete(paramName);
+      }
+    }
+
+    // Add provided parameters to the URL.
+    if (queryParams) {
+      for (const [paramName, decodedValue] of queryParams.entries()) {
+        unifiedParams = unifiedParams.set(paramName, decodedValue);
+      }
+    }
 
-    // Add requested parameters to the URL.
+    // Respect pre-existing params which already have values.
     for (const key of this.getAllKeys()) {
-      if (!excludeParams.has(key) && !existingParams.has(key)) {
+      if (
+        !excludeParams.has(key) &&
+        !urlParams.has(key) &&
+        !unifiedParams.has(key)
+      ) {
         const decodedValue = this.get(key);
         if (decodedValue) {
-          requestedParams = requestedParams.set(key, decodedValue);
+          unifiedParams = unifiedParams.append(key, decodedValue);
         }
       }
     }
 
     // Combine all requested parameters and their values, e.g. 'a=b'.
-    const joinedParams = requestedParams.toString();
+    const joinedParams = unifiedParams.toString();
+    const delimiter = root.includes('?') ? '&' : '?';
 
     // Build and return the final URL.
-    const [beforeFragment, fragment] = url.split('#', 2);
-    const delimiter = url.includes('?') ? '&' : '?';
+    return (
+      root +
+      (joinedParams.length === 0 ? '' : `${delimiter}${joinedParams}`) +
+      (fragment ? `#${fragment}` : '')
+    );
+  }
 
-    return joinedParams.length === 0
-      ? url
-      : `${beforeFragment}${delimiter}${joinedParams}` +
-          (fragment ? `#${fragment}` : '');
+  /**
+   * Removes undefined and null values from the given object and converts to a map.
+   */
+  #convertToCleanMap(
+    obj: Record<string, string | number | boolean | null | undefined>
+  ): Map<string, string | number | boolean> {
+    for (const key in obj) {
+      const value = obj[key];
+
+      if (value === undefined || value === null) {
+        delete obj[key];
```

Comment:
Deleting a property on this object is going to alter the original object which might have adverse downstream effects to the caller. If you really need to use a `Map`, I'd do that step first, then loop through the `Map` and delete entries that are `null` or `undefined`. Otherwise you could destructure it into a new object:

```
obj = {...obj};
```

Reply:
Oh, good catch. I'll clone the object before messing with it.

Reply:
Ended up removing this function altogether.